src/Interpreters/Streaming/Aggregator.cpp

```diff
@@ -103,6 +107,73 @@ Columns materializeKeyColumns(Columns & columns, ColumnRawPtrs & key_columns, co
 
     return materialized_columns;
 }
+
+Arena * getArena(AggregatedDataVariants & variants, AggregateStateType type)
+{
+    if (type == AggregateStateType::OnlyRetracted)
+        return variants.retracted_pool.get();
+    else
+        return variants.aggregates_pool;
+}
+
+template <typename BucketConverter>
+BlocksList concurrentBucketConvert(ThreadPool * thread_pool, const std::vector<Int64> & buckets, Arena * arena, Arenas & pools, BucketConverter && bucket_converter)
+{
+    std::atomic<UInt32> next_bucket_idx_to_merge = 0;
+    auto converter = [&](Arena * pool, ThreadGroupStatusPtr thread_group, const std::atomic_flag * cancelled) {
+        SCOPE_EXIT_SAFE(
+            if (thread_group)
+                CurrentThread::detachQueryIfNotDetached();
+        );
+
+        if (thread_group)
+            CurrentThread::attachToIfDetached(thread_group);
+
+        BlocksList blocks;
+        while (true)
+        {
+            if (cancelled && cancelled->test())
+                break;
+
+            UInt32 bucket_idx = next_bucket_idx_to_merge.fetch_add(1);
+            if (bucket_idx >= buckets.size())
+                break;
+
+            auto bucket = buckets[bucket_idx];
+            blocks.splice(blocks.end(), bucket_converter(bucket, pool));
+        }
+        return blocks;
+    };
+
+    size_t num_threads = thread_pool ? std::min(thread_pool->getMaxThreads(), buckets.size()) : 1;
+    if (num_threads <= 1)
+        return converter(arena, nullptr, nullptr);
+
+    /// Process in parallel
+    for (size_t i = pools.size(); i < num_threads; ++i)
+        pools.push_back(std::make_shared<Arena>());
```

Comment:
Do we need keep the arenas around which are used to parallel converting ? Or we can GC them after we are done ?

Reply:
So far I haven't found any use of arenas during the conversion process, meaning the new arena doesn't require GC and a temporary replacement can be used

Reply:
will use temp arenas